File: src/routes/assistant.ts
import { Router } from "express";
import { z } from "zod";
import { db } from "../db";
import { transactions } from "../db/schema";
import { eq, and, gte, lte, ilike, desc } from "drizzle-orm";
import { requireAuth, type AuthRequest } from "../middleware/auth";
import { classifyIntent } from "../services/nlg/templates";
import { analyticsEngine } from "../services/analytics/engine";
import type { NormalizedTransaction } from "../types";

const router = Router();

const fmt = (n: number) =>
  `₦${new Intl.NumberFormat("en-NG").format(Math.round(n))}`;

function toNormalized(row: typeof transactions.$inferSelect): NormalizedTransaction {
  return {
    date: row.date,
    amount: parseFloat(String(row.amount)),
    currency: row.currency,
    type: row.type as "debit" | "credit" | "transfer",
    description: row.description,
    rawDescription: row.rawDescription,
    merchant: row.merchant || undefined,
    entity: undefined,
    category: undefined,
    subcategory: row.subcategory || undefined,
    isRecurring: row.isRecurring,
    isSubscription: row.isSubscription,
    isBankCharge: row.isBankCharge,
    isDuplicate: row.isDuplicate,
  };
}

// POST /api/assistant/ask
router.post("/ask", requireAuth, async (req: AuthRequest, res, next) => {
  try {
    const Schema = z.object({
      question: z.string().min(1).max(500),
      dateRange: z
        .object({
          from: z.string().optional(),
          to: z.string().optional(),
        })
        .optional(),
    });

    const body = Schema.parse(req.body);
    const { question, dateRange } = body;

    const intent = classifyIntent(question);
    const now = new Date();
    const from = dateRange?.from
      ? new Date(dateRange.from)
      : new Date(now.getFullYear(), now.getMonth(), 1);
    const to = dateRange?.to ? new Date(dateRange.to) : now;

    let answer = "";
    let data: Record<string, unknown> = {};

    const baseConditions = [
      eq(transactions.userId, req.user!.userId),
      eq(transactions.isDuplicate, false),
      gte(transactions.date, from),
      lte(transactions.date, to),
    ];

    switch (intent.intent) {
      case "TOTAL_SPENDING": {
        const rows = await db
          .select()
          .from(transactions)
          .where(and(...baseConditions, eq(transactions.type, "debit")));
        const total = rows.reduce(
          (s, t) => s + parseFloat(String(t.amount)),
          0
        );
        answer = `You spent a total of ${fmt(total)} between ${from.toLocaleDateString("en-NG")} and ${to.toLocaleDateString("en-NG")} across ${rows.length} transactions.`;
        data = { total, count: rows.length };
        break;
      }

      case "INCOME_TOTAL": {
        const rows = await db
          .select()
          .from(transactions)
          .where(and(...baseConditions, eq(transactions.type, "credit")));
        const total = rows.reduce(
          (s, t) => s + parseFloat(String(t.amount)),
          0
        );
        answer = `You received ${fmt(total)} in income/credits during this period across ${rows.length} transactions.`;
        data = { total, count: rows.length };
        break;
      }

      case "CATEGORY_SPENDING": {
        const rows = await db
          .select()
          .from(transactions)
          .where(and(...baseConditions, eq(transactions.type, "debit")));
        const normalized = rows.map(toNormalized);
        const analytics = analyticsEngine.compute(normalized);
        const cat = analytics.byCategory.find(
          (c) =>
            c.category.toLowerCase() === intent.category?.toLowerCase()
        );
        if (cat) {
          answer = `You spent ${fmt(cat.amount)} on ${cat.category} during this period — that's ${Math.round(cat.percentage)}% of your total spending.`;
          data = cat;
        } else {
          answer = `No significant ${intent.category || "category"} spending found in this period.`;
        }
        break;
      }

      case "PERSON_SPENDING": {
        const name = intent.entity || "";
        const rows = await db
          .select()
          .from(transactions)
          .where(
            and(
              ...baseConditions,
              ilike(transactions.description, `%${name}%`)
            )
          );
        const total = rows
          .filter((r) => r.type === "debit")
          .reduce((s, t) => s + parseFloat(String(t.amount)), 0);
        const received = rows
          .filter((r) => r.type === "credit")
          .reduce((s, t) => s + parseFloat(String(t.amount)), 0);
        answer =
          total > 0
            ? `You sent ${fmt(total)} to ${name} across ${rows.filter((r) => r.type === "debit").length} transaction(s) this period.${received > 0 ? ` They also sent you ${fmt(received)}.` : ""}`
            : `No transactions found for "${name}" in this period.`;
        data = { sent: total, received, count: rows.length };
        break;
      }

      case "SUBSCRIPTION_LIST": {
        const rows = await db
          .select()
          .from(transactions)
          .where(and(...baseConditions, eq(transactions.isSubscription, true)));
        const total = rows.reduce(
          (s, t) => s + parseFloat(String(t.amount)),
          0
        );
        answer = `You have ${rows.length} subscription payment(s) totaling ${fmt(total)} during this period. Annually, that's approximately ${fmt(total * 12)} if the same subscriptions continue.`;
        data = {
          count: rows.length,
          total,
          items: rows.map((r) => ({
            merchant: r.merchant || r.description,
            amount: parseFloat(String(r.amount)),
          })),
        };
        break;
      }

      case "SAVINGS_OPPORTUNITIES": {
        const rows = await db
          .select()
          .from(transactions)
          .where(and(...baseConditions, eq(transactions.type, "debit")));
        const normalized = rows.map(toNormalized);
        const analytics = analyticsEngine.compute(normalized);
        const top = analytics.byCategory.slice(0, 3);
        const potentialSaving = top.reduce(
          (s, c) => s + c.amount * 0.2,
          0
        );
        answer = `Based on your spending pattern, your top categories are: ${top.map((c) => `${c.category} (${fmt(c.amount)})`).join(", ")}. If you reduced each by 20%, you could potentially save ${fmt(potentialSaving)}/month — that's ${fmt(potentialSaving * 12)}/year.`;
        data = { topCategories: top, potentialSaving };
        break;
      }

      case "LARGEST_TRANSACTION": {
        const [row] = await db
          .select()
          .from(transactions)
          .where(and(...baseConditions))
          .orderBy(desc(transactions.amount))
          .limit(1);
        if (row) {
          answer = `Your largest transaction was ${fmt(parseFloat(String(row.amount)))} — "${row.description}" on ${new Date(row.date).toLocaleDateString("en-NG")}.`;
          data = { amount: parseFloat(String(row.amount)), description: row.description, date: row.date };
        } else {
          answer = "No transactions found in this period.";
        }
        break;
      }

      case "COMPARE_PERIODS": {
        const prevFrom = new Date(from.getFullYear(), from.getMonth() - 1, 1);
        const prevTo = new Date(from.getFullYear(), from.getMonth(), 0);

        const [curr, prev] = await Promise.all([
          db.select().from(transactions).where(and(...baseConditions, eq(transactions.type, "debit"))),
          db.select().from(transactions).where(
            and(
              eq(transactions.userId, req.user!.userId),
              eq(transactions.isDuplicate, false),
              eq(transactions.type, "debit"),
              gte(transactions.date, prevFrom),
              lte(transactions.date, prevTo)
            )
          ),
        ]);

        const currTotal = curr.reduce((s, t) => s + parseFloat(String(t.amount)), 0);
        const prevTotal = prev.reduce((s, t) => s + parseFloat(String(t.amount)), 0);
        const change = prevTotal > 0 ? ((currTotal - prevTotal) / prevTotal) * 100 : 0;
        const dir = change > 0 ? "increased" : "decreased";

        answer = `This period spending: ${fmt(currTotal)}. Previous period: ${fmt(prevTotal)}. Spending ${dir} by ${Math.abs(Math.round(change))}%.`;
        data = { currTotal, prevTotal, changePercent: Math.round(change) };
        break;
      }

      default:
        answer =
          "I can answer questions like: 'How much did I spend on food?', 'How much did I send to Yuanna?', 'What are my subscriptions?', 'Compare this month with last month', or 'Where can I save money?'";
        data = { intent: "UNKNOWN", suggestions: true };
    }

    res.json({
      success: true,
      data: {
        question,
        answer,
        intent: intent.intent,
        context: data,
        period: {
          from: from.toISOString(),
          to: to.toISOString(),
        },
      },
    });
  } catch (err) {
    next(err);
  }
});

export default router;

File: app/routers/assistant.py
from datetime import date
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_, func
from app.database import get_db
from app.models.user import User
from app.models.transaction import Transaction
from app.middleware.auth import get_current_user
from app.middleware.casing import CamelModel
from app.services.nlg.templates import classify_intent, verbalize_savings_rate, verbalize_subscriptions
from app.services.analytics.engine import analytics_engine

router = APIRouter(prefix="/assistant", tags=["assistant"])


class AskRequest(CamelModel):
    question: str
    date_range: dict | None = None


@router.post("/ask", response_model=dict)
async def ask(
    body: AskRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    now = date.today()
    from_ = body.date_range.get("from") if body.date_range else None
    to_ = body.date_range.get("to") if body.date_range else None
    # When the user doesn't specify a period, answer over ALL their data — a
    # plain question shouldn't come back empty just because this calendar month
    # happens to have no transactions (e.g. a historical statement was uploaded).
    if from_ or to_:
        period_from = date.fromisoformat(from_) if from_ else date(2000, 1, 1)
        period_to = date.fromisoformat(to_) if to_ else now
    else:
        bounds = await db.execute(
            select(func.min(Transaction.date), func.max(Transaction.date)).where(
                and_(
                    Transaction.user_id == current_user.id,
                    Transaction.is_duplicate == False,
                )
            )
        )
        min_d, max_d = bounds.one()
        period_from = min_d or date(now.year, now.month, 1)
        period_to = max_d or now

    intent = classify_intent(body.question)
    tone = current_user.tone or "friendly"

    # Fetch transactions for period
    result = await db.execute(
        select(Transaction).where(
            and_(
                Transaction.user_id == current_user.id,
                Transaction.is_duplicate == False,
                Transaction.date >= period_from,
                Transaction.date <= period_to,
            )
        )
    )
    txs = result.scalars().all()
    tx_dicts = [
        {
            "date": t.date, "amount": float(t.amount), "type": t.type,
            "category": t.category, "subcategory": t.subcategory,
            "merchant": t.merchant, "description": t.description,
            "is_recurring": t.is_recurring, "is_subscription": t.is_subscription,
            "is_bank_charge": t.is_bank_charge, "is_duplicate": False,
            "entity_name": None,
        }
        for t in txs
    ]
    analytics = analytics_engine.compute(tx_dicts)

    answer = _answer(intent, body.question, analytics, tx_dicts, tone, period_from, period_to)
    return {"success": True, "data": {"answer": answer, "intent": intent}}


def _fmt(amount: float) -> str:
    return f"₦{amount:,.0f}"


def _answer(
    intent: str, question: str, analytics, txs: list[dict],
    tone: str, from_: date, to_: date,
) -> str:
    period_label = f"{from_.strftime('%b %d')} – {to_.strftime('%b %d, %Y')}"

    if intent == "TOTAL_SPENDING":
        return (
            f"You spent {_fmt(analytics.total_expenditure)} between {period_label} "
            f"across {len([t for t in txs if t['type'] == 'debit'])} transactions. "
            f"Your income was {_fmt(analytics.total_income)}, leaving {_fmt(analytics.net_cash_flow)} saved."
        )

    if intent.startswith("CATEGORY_SPENDING:"):
        cat = intent.split(":", 1)[1]
        stat = next((c for c in analytics.by_category if c.category.lower() == cat.lower()), None)
        if stat:
            return (
                f"You spent {_fmt(stat.total)} on {stat.category} ({period_label}). "
                f"That's {stat.percentage:.1f}% of your total spending and {stat.count} transactions."
            )
        return f"No {cat} spending found for {period_label}."

    if intent == "SUBSCRIPTION_LIST":
        sub_total = analytics.subscription_total
        sub_count = len([t for t in txs if t.get("is_subscription")])
        return verbalize_subscriptions(sub_total, sub_total * 12, sub_count, tone)

    if intent == "PERSON_SPENDING":
        if not analytics.by_entity:
            return "I didn't find any person-to-person transfers in this period."
        top = analytics.by_entity[:3]
        lines = [f"{e.name}: {_fmt(e.sent_total)} sent" for e in top if e.sent_total > 0]
        if not lines:
            return "No personal transfers found in this period."
        return "Top people you sent money to:\n" + "\n".join(f"• {l}" for l in lines)

    if intent == "SAVINGS_OPPORTUNITIES":
        recs = []
        if analytics.subscription_total > 0:
            recs.append(f"Subscriptions ({_fmt(analytics.subscription_total)}/mo) — review unused ones")
        if analytics.bank_charges_total > 0:
            recs.append(f"Bank charges ({_fmt(analytics.bank_charges_total)}) — consider a zero-fee digital bank")
        if analytics.by_category:
            top_cat = analytics.by_category[0]
            if top_cat.percentage > 25:
                saving = top_cat.total * 0.25
                recs.append(f"{top_cat.category} ({_fmt(top_cat.total)}) — a 25% cut saves {_fmt(saving)}/month")
        if not recs:
            return "Your spending looks well-controlled this period. Keep it up!"
        return "Potential savings opportunities:\n" + "\n".join(f"• {r}" for r in recs)

    if intent == "LARGEST_TRANSACTION":
        if analytics.largest_debit:
            tx = analytics.largest_debit
            return (
                f"Your largest spend was {_fmt(float(tx.get('amount', 0)))} — "
                f"{tx.get('description', 'Unknown')[:60]} on {tx.get('date', '')}."
            )
        return "No transactions found for this period."

    if intent == "INCOME_TOTAL":
        return (
            f"Your total income for {period_label} was {_fmt(analytics.total_income)}. "
            + verbalize_savings_rate(analytics.savings_rate, tone)
        )

    if intent == "COMPARE_PERIODS":
        return (
            f"This period ({period_label}): income {_fmt(analytics.total_income)}, "
            f"spent {_fmt(analytics.total_expenditure)}, saved {analytics.savings_rate:.1f}%. "
            f"For a detailed comparison, I need two periods — try specifying dates."
        )

    return (
        f"I can see {len(txs)} transactions for {period_label}. "
        f"Total income: {_fmt(analytics.total_income)}, total spent: {_fmt(analytics.total_expenditure)}. "
        f"Try asking me about a specific category, person, or subscription."
    )
